### PR TITLE
Functionality to discard changes

### PR DESCRIPTION
**Pull request by StanleyGoldman**
Fixes: #356

- Added functionality to discard all files or discard a list of files to `GitClient` using a new `GitCheckoutTask`
- Added functionality to `TreeBase` to return an array of child nodes from any given folder node
- Added a context menu to `ChangesView` to engage the discard functionality

### CODE REVIEW

**Comment by StanleyGoldman** [comment]
@mpOzelot when you get a chance try this out.

File: src/GitHub.Api/Git/RepositoryManager.cs [hunk 1]
```diff
@@ -283,6 +285,60 @@ public void UpdateGitStatus()
...
+            {
+                task = deleteItemsTask;
+            }
+            else if (gitDiscardTask != null)
+            {
+                task = gitDiscardTask;
+            }
+            else
+            {
+                throw new NotImplementedException();
```
Comment (StanleyGoldman) [inline src/GitHub.Api/Git/RepositoryManager.cs]: Should I throw an error? Or return an empty task here?

**Comment by lizenas** [comment]
Oh please let this be merged... Not being able to discard changes is driving me crazy.

**Comment by shana** [comment]
@lizenas Almost, there, just need to run some tests!

**Comment by StanleyGoldman** [comment]
@shana I'm seeing an error as a result of the changes in `a529043`.

```
180220-12:44:36.613 TRACE [ 4] <TaskManager>                       Exception on concurrent thread: 2 Finally
System.InvalidCastException: Cannot cast from source type to destination type.
  at <0x00000> <unknown method>
  at GitHub.Unity.ActionTask`1+<ActionTask>c__AnonStorey0[GitHub.Unity.GitStatusEntry[]].<>m__0 () [0x00093] in C:\Users\Spade\Projects\GitHub\Unity\src\GitHub.Api\Tasks\ActionTask.cs:90 
  at System.Threading.Tasks.Task.InnerInvoke () [0x00000] in <filename unknown>:0 
  at System.Threading.Tasks.Task.Execute () [0x00000] in <filename unknown>:0 
=======
   at GitHub.Unity.TaskManager+<ScheduleConcurrent>c__AnonStorey2`1[[GitHub.Unity.TaskBase, GitHub.Api, Version=0.27.0.0, Culture=neutral, PublicKeyToken=null]].<>m__0(System.Threading.Tasks.Task tt) in C:\Users\Spade\Projects\GitHub\Unity\src\GitHub.Api\Tasks\TaskManager.cs:line 146
   at System.Threading.Tasks.Task+<>c__DisplayClass6.<ContinueWith>b__5(System.Object obj)
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.Execute()
   at System.Threading.Tasks.Task.ExecutionContextCallback(System.Object obj)
   at System.Security.SecurityContext.Run(System.Security.SecurityContext securityContext, System.Threading.ContextCallback callback, System.Object state) in /Users/builduser/buildslave/mono/build/mcs/class/corlib/System.Security/SecurityContext.cs:line 150
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state) in /Users/builduser/buildslave/mono/build/mcs/class/corlib/System.Threading/ExecutionContext.cs:line 150
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef currentTaskSlot)
   at System.Threading.Tasks.Task.ExecuteEntry(Boolean bPreventDoubleExecution)
   at System.Threading.Tasks.TaskScheduler.TryExecuteTask(System.Threading.Tasks.Task task)
   at GitHub.Unity.ConcurrentExclusiveInterleave+ConcurrentExclusiveTaskScheduler.ExecuteTask(System.Threading.Tasks.Task task) in C:\Users\Spade\Projects\GitHub\Unity\src\GitHub.Api\Tasks\ConcurrentExclusiveInterleave.cs:line 398
   at GitHub.Unity.ConcurrentExclusiveInterleave.ExecuteConcurrentTask(System.Threading.Tasks.Task task) in C:\Users\Spade\Projects\GitHub\Unity\src\GitHub.Api\Tasks\ConcurrentExclusiveInterleave.cs:line 214
   at System.Threading.Tasks.Parallel+<>c__DisplayClass32`2[[System.Threading.Tasks.Task, System.Threading, Version=1.0.3333.0, Culture=neutral, PublicKeyToken=402899b480e6f383],[System.Object, mscorlib, Version=2.0.0.0, Culture=neutral, PublicKeyToken=b77a5c561934e089]].<PartitionerForEachWorker>b__30()
   at System.Threading.Tasks.Task.InnerInvoke()
   at System.Threading.Tasks.Task.InnerInvokeWithArg(System.Threading.Tasks.Task childTask)
   at System.Threading.Tasks.Task+<>c__DisplayClass3.<ExecuteSelfReplicating>b__2(System.Object )
   at System.Threading.Tasks.Task.ExecuteSelfReplicating(System.Threading.Tasks.Task root)
   at System.Threading.Tasks.Task.Execute()
   at System.Threading.Tasks.Task.ExecutionContextCallback(System.Object obj)
   at System.Security.SecurityContext.Run(System.Security.SecurityContext securityContext, System.Threading.ContextCallback callback, System.Object state) in /Users/builduser/buildslave/mono/build/mcs/class/corlib/System.Security/SecurityContext.cs:line 150
   at System.Threading.ExecutionContext.Run(System.Threading.ExecutionContext executionContext, System.Threading.ContextCallback callback, System.Object state) in /Users/builduser/buildslave/mono/build/mcs/class/corlib/System.Threading/ExecutionContext.cs:line 150
   at System.Threading.Tasks.Task.ExecuteWithThreadLocal(System.Threading.Tasks.Task ByRef currentTaskSlot)
   at System.Threading.Tasks.Task.ExecuteEntry(Boolean bPreventDoubleExecution)
   at System.Reflection.MonoMethod.InternalInvoke(System.Object , System.Object[] , System.Exception ByRef )
   at System.Reflection.MonoMethod.Invoke(System.Object obj, BindingFlags invokeAttr, System.Reflection.Binder binder, System.Object[] parameters, System.Globalization.CultureInfo culture) in /Users/builduser/buildslave/mono/build/mcs/class/corlib/System.Reflection/MonoMethod.cs:line 222
   at System.Reflection.MethodBase.Invoke(System.Object obj, System.Object[] parameters) in /Users/builduser/buildslave/mono/build/mcs/class/corlib/System.Reflection/MethodBase.cs:line 115
   at GitHub.Unity.TaskSchedulerExcludingThread.ExecuteEntry(System.Threading.Tasks.Task task, Boolean flag) in C:\Users\Spade\Projects\GitHub\Unity\src\GitHub.Api\Tasks\ConcurrentExclusiveInterleave.cs:line 33
   at GitHub.Unity.TaskSchedulerExcludingThread.TaskExecuteWaitCallback(System.Object obj) in C:\Users\Spade\Projects\GitHub\Unity\src\GitHub.Api\Tasks\ConcurrentExclusiveInterleave.cs:line 49
```

File: src/GitHub.Api/Git/RepositoryManager.cs [hunk 2]
```diff
@@ -301,51 +301,41 @@ public ITask DiscardChanges(GitStatusEntry[] gitStatusEntries)
...
+                    }
+
+                    ITask<string> gitDiscardTask = null;
+                    if (itemsToRevert.Any())
+                    {
+                        gitDiscardTask = GitClient.Discard(itemsToRevert);
+                        task.Then(gitDiscardTask);
+                    }
+                }
+                , () => gitStatusEntries);
```
Comment (StanleyGoldman) [inline src/GitHub.Api/Git/RepositoryManager.cs]: I'm not sure what about this change causes a problem.

**Comment by StanleyGoldman** [comment]
Also as a result of #541, what is the correct functionality when running discard on a asset with a child `.meta`?

image

**Comment by shana** [comment]
> Also as a result of #541, what is the correct functionality when running discard on a asset with a child .meta?

Probably discard the changes to the meta file as well. If we don't, Unity will detect the change to the parent and recreate the meta file accordingly, which might or might not result in the meta file changes going away (if Unity serializes the parent after the discard in a slightly different order, it might end up with a meta file that has no data changes but still looks changes). We might as well do that ourselves.

**Comment by StanleyGoldman** [comment]
@shana I tried debugging that issue but I didn't have much luck with it.

**Comment by shana** [comment]
Ok so the crasher in this PR is being handled in #609 #604 and #608, and we kinda need all of them to make everything work properly, so I'm landing this one with the note that it's incomplete

**Pull request merged**